Introduce multistore technical documentation for configuration forms and settings (#843)
PrestaShop/docs: src/content/1.7/development/multistore/configuration-forms/_index.md

---
title: Configuration forms
menuTitle: Configuration forms
summary: "Configuration forms when multistore is enabled"
weight: 50
---

# Configuration forms in multistore context
{{< minver v="1.7.8" title="true" >}}

> 

Reviewer: ```suggestion

The Back Office contains multiple configuration forms that control settings of the shop. When multistore is enabled, Back Office user must be able to modify these settings either for all shops, for a group of shop or for a single shop.

```

Since Prestashop 1.7.8, when multistore is enabled, migrated configuration forms can display a checkbox before fields in order to specify if the configuration value is set for the current context shop/group, or if it is a value inherited from all contexts, or from a shop group.

> If you're developing a configuration form and you want it to take into account the current multistore context, then you must make your form compatible with multistore checkboxes. This is made possible thanks to [Symfony's form extensions mechanism](https://symfony.com/doc/3.4/form/create_form_type_extension.html).

Reviewer: ```suggestion
If you are building a configuration form and you want the form to be aware of the current multistore context, then you must make your form compatible with multistore checkboxes. This is made possible thanks to [Symfony's form extensions mechanism](https://symfony.com/doc/3.4/form/create_form_type_extension.html).
```

```php
<?php

    /**
     * {@inheritdoc}
     *
     * @see MultistoreConfigurationTypeExtension
     */
    public function getParent(): string
    {
        return MultistoreConfigurationType::class;
    }
```

Then for each form elements that should have its multistore checkbox, you must add a `multistore_configuration_key` attribute, with the configuration key to which it should be attached as a value:

```php
<?php
class MyFormType extends TranslatorAwareType
{
    /**
     * {@inheritdoc}
     */
    public function buildForm(FormBuilderInterface $builder, array $options)
    {
        $builder
            ->add(
                'my_field',
                SwitchType::class,
                [
                    'required' => true,
                    // Here we add the multistore_configuration_key attribute with the configuration key it is linked to.
                    'attr' => [
                        'multistore_configuration_key' => 'THE_CONFIGURATION_KEY',
                    ],
                    // ...
                ]
            )
        // ...

```

> Last but not least, you must make the frontend aware of these state modifications, you just need to add the `MultistoreConfigField` in your javascript:

Reviewer: ```suggestion
Last but not least, you must make the frontend aware of these state modifications, you need to add the `MultistoreConfigField` in your javascript:
```
May I suggest to avoid words like "just", "simply", "easy" ? 😉 

This is because of https://www.knowledgeowl.com/home/dont-say-simply-jim-fisher

- Thanks to the javascript component, checking or unchecking a checkbox will enable/disable the field for the current shop context.

## Saving data from a multistore configuration form

When receiving the data from the form, you will also receive values from the multistore checkboxes that were added to your form:

- When a multistore checkbox is pushed unchecked, we want to delete the configuration value for the current context (so that it inherits its value from parent contexts).
- When a multistore checkbox is pushed checked, we want to save the configuration value for the current context.

When your configuration form is multistore compatible, your configuration class should extend the `AbstractMultistoreConfiguration` abstract class instead of directly implementing the `DataConfigurationInterface` interface. This will give you access to helper methods that will make it easy to store the right configuration values at the right place and for the right context.

The two helper methods are

- **`getShopConstraint()`:** this will give you a ShopConstraint object reflecting the current context
```php
public function getShopConstraint(): ?ShopConstraint
```

> - **`updateConfigurationValue()`:** this will save or delete the configuration value for the given configuration key, taking into account the state of the multistore checkbox and the current shop context. Note that this helper method is totally compatible with non multistore fields.

Reviewer: ```suggestion
- **`updateConfigurationValue()`:** this will save or delete the configuration value for the given configuration key, and is aware of the multistore checkbox and the current shop context. Note that this helper method is fully compatible with not-multistore fields.
```

```php
public function updateConfigurationValue(string $configurationKey, string $fieldName, array $input, ?ShopConstraint $shopConstraint, array $options = []): void
```

> As an example, see how it's used in the `MaintenanceConfiguration` class

Reviewer: ```suggestion
As an example, see how it is used in the `MaintenanceConfiguration` class
```
I like to consider we should look for a "formal" tone in devdocs, so no abbreviations. But this depends on people I suppose 😄  🤔 📘 .
